base/exporter.go

```diff
@@ -0,0 +1,318 @@
+}
+
+type QueryStat struct {
+	QueryCount      *SgwIntStat
+	QueryErrorCount *SgwIntStat
+	QueryTime       *SgwIntStat
+}
+
+func (s *SgwStats) DBStats(name string) *DbStats {
+
+	if s.DbStats == nil {
+		s.DbStats = map[string]*DbStats{}
+	}
+
+	if _, ok := s.DbStats[name]; !ok {
+		s.DbStats[name] = &DbStats{}
+	}
+	return s.DbStats[name]
+}
+
+func (d *DbStats) DeltaSync(dbName string) *DeltaSyncStats {
+	if d.DeltaSyncStats == nil {
+		d.DeltaSyncStats = &DeltaSyncStats{
+			DeltasRequested:           NewIntStat("delta_sync", "deltas_requested", dbName, prometheus.CounterValue, 0),
+			DeltasSent:                NewIntStat("delta_sync", "deltas_sent", dbName, prometheus.CounterValue, 0),
+			DeltaPullReplicationCount: NewIntStat("delta_sync", "delta_pull_replication_count", dbName, prometheus.CounterValue, 0),
+			DeltaCacheHit:             NewIntStat("delta_sync", "delta_cache_hit", dbName, prometheus.CounterValue, 0),
+			DeltaCacheMiss:            NewIntStat("delta_sync", "delta_sync_miss", dbName, prometheus.CounterValue, 0),
+			DeltaPushDocCount:         NewIntStat("delta_sync", "delta_push_doc_count", dbName, prometheus.CounterValue, 0),
+		}
+	}
+	return d.DeltaSyncStats
+}
+
+func (d *DbStats) SharedBucketImport() *SharedBucketImportStats {
+	if d.SharedBucketImportStats == nil {
+		dbName := d.dbName
+		d.SharedBucketImportStats = &SharedBucketImportStats{
+			ImportCount:          NewIntStat("shared_bucket_import", "import_count", dbName, prometheus.CounterValue, 0),
+			ImportCancelCAS:      NewIntStat("shared_bucket_import", "import_cancel_cas", dbName, prometheus.CounterValue, 0),
+			ImportErrorCount:     NewIntStat("shared_bucket_import", "import_error_count", dbName, prometheus.CounterValue, 0),
+			ImportProcessingTime: NewIntStat("shared_bucket_import", "import_processing_time", dbName, prometheus.GaugeValue, 0),
+			ImportHighSeq:        NewIntStat("shared_bucket_import", "import_high_seq", dbName, prometheus.CounterValue, 0),
+			ImportPartitions:     NewIntStat("shared_bucket_import", "import_partitions", dbName, prometheus.GaugeValue, 0),
+		}
+	}
+
+	return d.SharedBucketImportStats
+}
+
+func (d *DbStats) GSIStats(queryName string) *QueryStat {
+	if d.GsiStats == nil {
+		d.GsiStats = &GsiStats{
+			Stats: map[string]*QueryStat{},
+		}
+	}
+
+	if _, ok := d.GsiStats.Stats[queryName]; !ok {
+		dbName := d.dbName
+		d.GsiStats.Stats[queryName] = &QueryStat{
+			QueryCount:      NewIntStat("gsi_views", queryName+"_query_count", dbName, prometheus.CounterValue, 0),
+			QueryErrorCount: NewIntStat("gsi_views", queryName+"_query_error_count", dbName, prometheus.CounterValue, 0),
+			QueryTime:       NewIntStat("gsi_views", queryName+"_query_time", dbName, prometheus.CounterValue, 0),
+		}
+	}
+
+	return d.GsiStats.Stats[queryName]
+}
+
+func (g *GsiStats) MarshalJSON() ([]byte, error) {
+	ret := map[string]interface{}{}
+	for queryName, queryMap := range g.Stats {
+		ret[queryName+"_query_count"] = queryMap.QueryCount
```

Comment:
These names seem like they don't align with the existing names (e.g. StatKeyN1qlQueryCountExpvarFormat and StatKeyViewQueryCountExpvarFormat) - I think the _query_ should be removed?

Reply:
You might be right. I haven't done a careful pass over checking that all the names are the same as they are now. I'll make sure I check over these a bunch before I submit for a full review